### PR TITLE
Add option to extract opaque bodies

### PR DESCRIPTION
**Pull request by Nadrieril**
With `--extract-opaque-bodies`, we now also extract foreign globals, methods, trait provided methods, and ADT definitions. This is limited by two things:
- not all function bodies have a MIR in the crate data (only inlineable and generic functions do);
- when function have MIR available it's only optimized MIR, which we don't support as well.

This PR includes a few tests that fail because of other problems. Two of them are hax limitations that I'm ignoring until we're back on hax master. I opened issues for the other two: #123, #120.

The PR also includes many improvements to the ui tests that I needed along the way.

Fixes https://github.com/AeneasVerif/charon/issues/114

### CODE REVIEW

File: charon/src/translate_traits.rs [hunk 1]
```diff
@@ -304,7 +304,7 @@ impl<'tcx, 'ctx> TransCtx<'tcx, 'ctx> {
                     // but still remember their name.
```
Comment (sonmarcho) [inline charon/src/translate_traits.rs:304]: Update the comment to mention the CLI option which changes this behavior?

File: charon/src/translate_functions_to_ullbc.rs [hunk 2]
```diff
@@ -1887,14 +1883,12 @@ impl<'tcx, 'ctx> TransCtx<'tcx, 'ctx> {
...
 
-        let body = if rust_id.is_local() && is_transparent {
-            // It's a local and transparent global: we extract its body as for functions.
-            match bt_ctx.translate_body(rust_id.expect_local(), 0) {
-                Err(_) => {
-                    // Error case: we could have a specific variant
-                    None
-                }
-                Ok(body) => Some(body),
+        let body = if is_transparent {
```
Comment (sonmarcho) [inline charon/src/translate_functions_to_ullbc.rs:1886]: Maybe add a comment here to explain that we test whether the definition is local or not in `translate_body`(I was quite puzzled when reviewing this change at first)

File: charon/src/translate_functions_to_ullbc.rs [hunk 3]
```diff
@@ -1429,16 +1429,6 @@ impl<'tcx, 'ctx, 'ctx1> BodyTransCtx<'tcx, 'ctx, 'ctx1> {
...
         for arg in args {
-            // There should only be moved arguments, or constants
-            match arg {
-                hax::Operand::Move(_) | hax::Operand::Constant(_) => {
-                    // OK
-                }
-                hax::Operand::Copy(_) => {
-                    unreachable!();
-                }
-            }
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply charon/src/translate_functions_to_ullbc.rs:1440]: Yep, in one of the std functions exported by one of the tests, I didn't find out which one exactly

File: charon/tests/ui/issue-114-opaque-bodies-aux.rs [hunk 4]
```diff
@@ -0,0 +1,21 @@
+//@ skip
```
[the remark this answers is not in the document]
Comment (Nadrieril) [reply charon/tests/ui/issue-114-opaque-bodies-aux.rs:1]: Yep

File: charon/src/translate_functions_to_ullbc.rs (hunk 2 above)
Comment (Nadrieril) [reply charon/src/translate_functions_to_ullbc.rs:1886]: We could do the `is_transparent` check in `translate_body` instead?

**Comment by Nadrieril** [comment]
We could pass `-Zalways-encode-mir` to dependencies, I think this ensures we get the MIR for everything (while we're at it, `-Zno-codegen` avoids generating code we don't need).

This doesn't work for `std` though; for that, this proposal might be the solution, as it proposes shipping a std with full MIR: https://github.com/rust-lang/compiler-team/issues/738. Or we could ship our own MIR-only sysroot.

File: charon/src/translate_functions_to_ullbc.rs (hunk 2 above)
Comment (Nadrieril) [reply charon/src/translate_functions_to_ullbc.rs]: I did that, lmk if you like it

**Comment by zhassan-aws** [comment]
> The fact that we can't always retrieve the bodies of the external functions makes the effects of this PR quite limited, unfortunately. I wonder how the Kani people do, as I thought Kani looked into the bodies of the external definitions (@zhassan-aws could you comment on that?).

In Kani, we run the `kani-compiler` on each of the dependent crates to generate the MIR only. Then, when compiling the target user crate, the `kani-compiler` has access to the MIR of all the dependent crates. This also applies to `std`, but it's handled in a special way.

See [this RFC](https://model-checking.github.io/kani/rfc/rfcs/0001-mir-linker.html) and [one of the PRs that implements it](https://github.com/model-checking/kani/pull/1683) for more information.

Feel free to reach out for further information.

**Comment by Nadrieril** [comment]
Thank you for these links!

**Comment by sonmarcho** [comment]
This is useful documentation, thanks!

**Review by sonmarcho — APPROVED** [review]
Looks good: it indeed makes sense to do check once and for all in `translate_body` whether we should extract the body or not.

**Pull request merged**